### PR TITLE
LPD-34718 Add "Parameters" and "URL Preview" inputs on Details tab

### PR DESCRIPTION
**Pull request by thektan**
Story: https://liferay.atlassian.net/browse/LPD-25241
Subtask: https://liferay.atlassian.net/browse/LPD-34718

- Adds a parameters input to add additional parameters to the request
- Behind feature flag LPD-25230 
- Documentation link is also added but it'll only appear once a URL is added to the [frontend-data-set-admin-web.json](https://github.com/liferay-frontend/liferay-portal/pull/4405/commits/17d24163e9bb8257e40cdc102d525e14b0e551c8#diff-f61567cf110ad80b73fb39cd1b38e0284880491638380ca4fcae4e0676e54bd7) and is published on the hosted server.

Screenshot 2024-09-11 at 5 40 21 PM

Added to the end of the URL:

Screenshot 2024-09-05 at 8 43 26 PM

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java [hunk 1]
```diff
@@ -304,6 +304,8 @@ private String _buildFragmentHTML(
 				"namespace", fragmentRendererContext.getFragmentElementId()
 			).put(
 				"pagination", _getPaginationJSONObject(fdsViewObjectEntry)
+			).put(
+				"parameters", fdsViewObjectEntry.getPropertyValue("parameters")
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java:308]: Overall approach looks good to me. I'm a bit concerned with the naming, in particular, `parameter` looks to broad. To begin, the endpoint URL may already contain something we could misinterpret as parameter in the path, e.g. `/v1.0/user-accounts/{userAccountId}/organizations`, which are totally unrelated to the feature we're introducing here.

Also, there is the idea that those parameters are not the only ones that will end up being added to the final URL used by FDS to fetch data. Pagination, filter, nested fields are dynamically added by the FDS/Fragment when needed.

Having said that, I'd like to start a short discussion about what should be the ideal name for this.

Considering we're storing it as string, a reasonable name to me would contain mentions to "query", "string" and maybe "endpoint", to make it more accurate. It might be useful to highlight these ones are _additional_ to signal they're not exclusive (implementation will include them but others too)

Something along the lines of `enpdointQueryString`, `queryString`, `additionalQueryString` might work here.

Please note I'm talking from the DSM/FDS api perspective. UI can keep the name product/design feels best for the end user intended purpose. @thektan @herrardo

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js [hunk 2]
```diff
@@ -156,6 +157,16 @@ export async function loadData(
 		);
 	}
 
+	if (Liferay.FeatureFlags['LPD-25230'] && parameters) {
+		const parametersArray = parameters.split('&');
+
+		parametersArray.forEach((parameter) => {
+			const [key, value] = parameter.split('=');
+
+			url.searchParams.append(key, value);
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js:166]: A small nit: we need to ensure that `append()` is enough for our purposes. In other words, check there's no need to "merge" values of a parameter both managed by the FDS/DSM (like `nestedFields`, see [here](https://github.com/liferay/liferay-portal/blob/3954fa615adb3574dfa76dec7bc962da5d2d36c1/modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java#L861-L869)) and also included in the list of parameters manually added.

The use of `parameters` is likely to produce these "clashing" situation where FDS wants to set a filter but also parameter contain a `filter` already. 

By using `append()` we would include 2 separate `filter` query parameters, I'm not sure how endpoints will behave.

It's clear that this is an advanced setting, and thus we assume user "knows what (s)he is doing", however, if adding an initial filter to the fetched data prevents data set to include any other filter, the feature would look too limiting imho

@herrardo I'd say we could at least analyze current behavior for `filter` and `nestedFields`. Makes sense?

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java (hunk 1 above)
Comment (thektan) [reply modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java:308]: Ah, yes I agree with that. Those suggestions all sound good. 

I suppose a strange thing with this `parameters` is it feels like it should be part of the `apiURL`, but it needs to be passed in separately so it can be merged with the other configured `sorts`/`filters`.

So I'm thinking `additionalQueryString` so it's clear it's more of like an "add on".

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js (hunk 2 above)
Comment (thektan) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js]: I pushed up additional changes to merge the `sort`, `filter`, and `nestedFields` parameters.

For the filters I was trying to ignore the existing fields used, but it was very tricky because of the different ways a filter could be constructed. Plus I think it might cause very unexpected things by modifying the filter being attempted to pass in. I currently have the filter added on with an `and`. Do you see any issues with this approach?

https://github.com/liferay-frontend/liferay-portal/blob/093c11eff4249836bbe07e57a08761e3b2d6320d/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js#L174-L177

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java (hunk 1 above)
Comment (herrardo) [reply modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java:308]: I know that `parameters` is quite broad but is still accurate: https://support.google.com/google-ads/answer/6277564?hl=en not that they use 'URLParameters'

It's true that there are another names such as `queryString` or even the `URLSearchParams` interface used in JS.

Also, we know they will other parameters (pagination, filtering and sorting coming from the DSM) but they are the same kind of parameters so i would name them in the same fashion, adding a prefix to the ones treated in this PR, so, possibilities:

- URLParameters  (the ones from DSM)
- URLExtraParameters (the ones in this PR)

- URLParameters  (the ones from DSM)
- URLStaticParameters (the ones in this PR)

- QueryString(the ones from DSM)
- AdditionalQueryString (the ones in this PR)

Just ideas to see is you like them ;)

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js (hunk 2 above)
Comment (herrardo) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js]: I am requesting using different filter combinations such as:

`(alternateName eq 'thomas') and (alternateName eq 'test')`

`(alternateName eq 'thomas') and (givenName eq 'test')`

`(alternateName eq 'thomas') and (givenName eq+'Thomas')`

to understand is applies correctly the inclusion/exclusion using the same field or different and seems to work just fine.

Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js]: I'm assuming there is no possibility to pass 2 different `filter` parameters in the URL and magically expect the endpoint to obey both of them as the filter expressions need to be connected with a logical operator which is missing.

In that line of thought, I guess we have no choice but attempt some sort of "logical merging" in the filter due to the requirements we have to solve. This would entail:

- Having one single filter expression all the time
- Connect the expression provided in the parameters to the filter expression set by the filters

Said that, the implementation details may vary. Please note we're already building an odata expression out from a  combined filter expression [including possible previous values for the filter parameter](https://github.com/liferay/liferay-portal/blob/56a7c977f0e1d64b668de3cb8e1e631aac1e975f/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js#L132-L137). So we might be reinventing the wheel here.

Anyways, this way of connecting is something we're already doing when more than 1 filter is applied at the same time so this looks like the way to go in this case.

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java (hunk 1 above)
Comment (thektan) [reply modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java:308]: Ended up going with `queryString` to keep it simple for a field name on the `FDSView` object as well.

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:test:relevant

**Comment by thektan** [comment]
Looking into the test failure.

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java]: So Dani pointed out that I should add the new field from this PR to [migration to JSON](https://github.com/liferay-frontend/liferay-portal/pull/4414). Without context of discussion here, I found `queryString` to be very confusing. We typically associate `query` with search, and `*String` seems unnecessary.

I'd call this `additionalURLParameters`.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js [hunk 3]
```diff
@@ -156,6 +157,98 @@ export async function loadData(
...
+					(queryStringSortValueItem) => {
+						if (
+							!existingSortParamFields.includes(
+								queryStringSortValueItem.split(':')[0]
+							)
+						) {
+							newSortParams.push(queryStringSortValueItem);
+						}
+					}
+				);
```
[the remark this answers is not in the document]
Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js:206]: Good question Marko. To me, this needs a bit of consideration. Because the original intent is to have "Additional parameters", it's not clear to me that those mean "merge" vs "override".

At the very least, I'd consider "filter" as a special case where "merge" is desirable. For the rest, we can just have an override. Let's talk about this early morning next week so that we can finish this.

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java (hunk 1 above)
Comment (thektan) [reply modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java]: This is good with me. I'll rename this to `additionalAPIURLParameters` in the meantime.

`queryString` made sense to me since it seems to be an accurate description of the value. For example, a wiki on [Query String](https://en.wikipedia.org/wiki/Query_string). There's also a popular js package called [`query-string`](https://github.com/sindresorhus/query-string) which interprets strings like `?foo=bar`.

That being said, `additionalAPIURLParameters` does seem appropriate in the context of DSM and FDS because of the naming used in the UI and how it's being combined with the `apiURL`.

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js [hunk 4]
```diff
@@ -148,14 +149,96 @@ export async function loadData(
...
+			const [key, value] = parameter.split('=');
+
+			const existingFilter = url.searchParams.get('filter');
+
+			if (key === 'filter' && existingFilter) {
+				url.searchParams.set(
+					'filter',
+					existingFilter + ' and ' + value
+				);
+			}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js:172]: We discussed this in a meet and the conclusion is that yes, there is a product need for the "merge" behavior. My concern is that this `' and '` may not be sufficient, because we have few different ways in building odata. See [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/controls/filters/implementation/SelectionFilter.tsx#L104-L139) and [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/controls/filters/implementation/DateRangeFilter.tsx#L82-L107):
- Collection entity field has strict format, enclosed in parentheses after keyword `any`, `/any(x:(x eq 1) and (x eq 2))`. I am not sure that appending `and` will work as expected. It might just break parsing, or have no effect. 
- Should we enclose `existingFilter` and `value` with parentheses? What would happen if exclude `or` is in `existingFilter`? We are at a mercy of order of operators.
- Client extension filter can have any format of odata, but I think we can safely ignore that case. There is no way to cover it.

Can we test all [entity field types](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/constants/FDSEntityFieldTypes.java)? Custom FDS module has example of all 3/4 entity field types, missing only `date-time`. To test if they work, I think you can temporarily hardcode to always add some additional parameters.

**Review by markocikos — COMMENTED** [review]
@thektan a comment inline

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js]: See https://github.com/liferay-frontend/liferay-portal/pull/4405#discussion_r1761269382 I wanted to highlight most relevant code.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js (hunk 4 above)
Comment (thektan) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js:172]: I agree that we can enclose with parentheses to keep the order of operators with a more expected outcome.

With a collection entity field, I'd have to figure out how to test this, but I'd assume if the `existingFilter` was [a lambda operator](https://help.liferay.com/hc/en-us/articles/360036343152-Filter-Sort-and-Search#lambda-operators) `keywords/any(k:contains(k,'substring1'))` and the new `value` was also similar `keywords/any(k:contains(k,'substring2')`, it should be fine combined as `(keywords/any(k:contains(k,'substring1'))) and (keywords/any(k:contains(k,'substring2')))`

I'm unclear what you mean by testing all entity field types? I think it'd be up to the user to format the value that'd be added on.

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/index.js:172]: I meant to manually test all entity field types, as a way to manually test all types of odata formats we have. But, it looks like we have it all covered, other entity types look to me like they do fit `(existing) and (new)`. If there is some odd edge case, we'll see it down the line. So, as far as I'm concerned, feel free to ship this, after you add the parenthesis.

**Comment by thektan** [comment]
ci:test:sf

**Comment by thektan** [comment]
ci:forward

**Comment by thektan** [comment]
Odd to see these unique failures after just adding parentheses. They don't seem related, but I'm going to try running it again.

**Comment by thektan** [comment]
ci:forward
